# IBM Data Science Captstone Project Python Notebook

We will use this notebook to complete Capstone project

In [1]:
import pandas as pd
import numpy as np

In [2]:
print('Hello Capstone Project Course!')

Hello Capstone Project Course!


# Requirement

- The dataframe will consist of three columns: PostalCode, Borough, and Neighborhood
- Only process the cells that have an assigned borough. Ignore cells with a borough that is Not assigned.
- More than one neighborhood can exist in one postal code area. For example, in the table on the Wikipedia page, you will notice that M5A is listed twice and has two neighborhoods: Harbourfront and Regent Park. These two rows will be combined into one row with the neighborhoods separated with a comma as shown in row 11  in the above table.
- If a cell has a borough but a Not assigned  neighborhood, then the neighborhood will be the same as the borough.
- Clean your Notebook and add Markdown cells to explain your work and any assumptions you are making.
- In the last cell of your notebook, use the .shape method to print the number of rows of your dataframe.

# Q1 - Shape of our data

In [3]:
import requests
from bs4 import BeautifulSoup

In [4]:
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
wiki_url = requests.get(url)
wiki_url = requests.get(url).text

In [5]:
soup=BeautifulSoup(wiki_url,"html.parser")

Lets generate dataframe from the data we scraped

In [84]:
table_contents=[]
table=soup.find('table')
for row in table.findAll('td'):
    cell = {}
    if row.span.text=='Not assigned':
        pass
    else:
        cell['PostalCode'] = row.p.text[:3]
        cell['Borough'] = (row.span.text).split('(')[0]
        cell['Neighbourhood'] = (((((row.span.text).split('(')[1]).strip(')')).replace(' /',',')).replace(')',' ')).strip(' ')
        table_contents.append(cell)

# print(table_contents)
df=pd.DataFrame(table_contents)
df['Borough']=df['Borough'].replace({'Downtown TorontoStn A PO Boxes25 The Esplanade':'Downtown Toronto Stn A',
                                             'East TorontoBusiness reply mail Processing Centre969 Eastern':'East Toronto Business',
                                             'EtobicokeNorthwest':'Etobicoke Northwest','East YorkEast Toronto':'East York/East Toronto',
                                             'MississaugaCanada Post Gateway Processing Centre':'Mississauga'})


In [85]:
df.head()

PostalCode           Borough                     Neighbourhood
0        M3A        North York                         Parkwoods
1        M4A        North York                  Victoria Village
2        M5A  Downtown Toronto         Regent Park, Harbourfront
3        M6A        North York  Lawrence Manor, Lawrence Heights
4        M7A      Queen's Park     Ontario Provincial Government

We will now group our data based on Postal Code

In [86]:
df = df.groupby(['PostalCode']).head()
df

PostalCode                Borough  \
0          M3A             North York   
1          M4A             North York   
2          M5A       Downtown Toronto   
3          M6A             North York   
4          M7A           Queen's Park   
..         ...                    ...   
98         M8X              Etobicoke   
99         M4Y       Downtown Toronto   
100        M7Y  East Toronto Business   
101        M8Y              Etobicoke   
102        M8Z              Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4                        Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100                                     Enclave of M4L  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [9]:
df.shape

(103, 3)

For 1st Section, we have 103 rows and 3 columns in our df 

# Q2 - Geocoder

In [12]:
#!pip install geocoder

In [13]:
import geocoder

# initialize your variable to None
lat_lng_coords = None
postal_code = 'M1B'

# loop until you get the coordinates
while(lat_lng_coords is None):
  g = geocoder.google('{}, Toronto, Ontario'.format(postal_code))
  lat_lng_coords = g.latlng

latitude = lat_lng_coords[0]
longitude = lat_lng_coords[1]

Worked on the above code however, it was taking ages to run so will use csv file which was provided 

In [14]:
data = pd.read_csv('Geospatial_Coordinates.csv')
data.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

Let's check the rows and columns of our datasets and df

In [15]:
print("The shape of df data is: ", df.shape)
print("The shape of data is: ", data.shape)

The shape of df data is:  (103, 3)
The shape of data is:  (103, 3)


Since both datasets have similar columns and rows, we can join our data, lets see how many common columns we have to join our dataset

In [16]:
print("Column names in df dataset is as follows: ",df.columns)
print("Column names in data is as follows: ", data.columns)

Column names in df dataset is as follows:  Index(['PostalCode', 'Borough', 'Neighborhood'], dtype='object')
Column names in data is as follows:  Index(['Postal Code', 'Latitude', 'Longitude'], dtype='object')


We can use Postal Code as a column to join our dataset

Since column name on data is Postal Code lets change it to PostalCode so it matches with df data

In [87]:
data.rename(columns={'Postal Code':'PostalCode'}, inplace=True)
data.head()

PostalCode   Latitude  Longitude
0        M1B  43.806686 -79.194353
1        M1C  43.784535 -79.160497
2        M1E  43.763573 -79.188711
3        M1G  43.770992 -79.216917
4        M1H  43.773136 -79.239476

In [88]:
combined_data = df.join(data.set_index('PostalCode'), on='PostalCode', how='inner')
combined_data

PostalCode                Borough  \
0          M3A             North York   
1          M4A             North York   
2          M5A       Downtown Toronto   
3          M6A             North York   
4          M7A           Queen's Park   
..         ...                    ...   
98         M8X              Etobicoke   
99         M4Y       Downtown Toronto   
100        M7Y  East Toronto Business   
101        M8Y              Etobicoke   
102        M8Z              Etobicoke   

                                         Neighbourhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
3                     Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4                        Ontario Provincial Government  43.662301 -79.389494  
..                                                 ...        ...        ...  
98       The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944  
99                                Church and Wellesley  43.665860 -79.383160  
100                                     Enclave of M4L  43.662744 -79.321558  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509  
102  Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999  

[103 rows x 5 columns]

We have sucessfully joined our df with data report

Lets check the shape of our new data

In [19]:
combined_data.shape

(103, 5)

## End of 2nd Question

----------------------------- #### --------------------------

# Quetion 3

Explore and cluster the neighborhoods in Toronto. You can decide to work with only boroughs that contain the word Toronto and then replicate the same analysis we did to the New York City data. It is up to you. 

Just make sure:

- to add enough Markdown cells to explain what you decided to do and to report any observations you make. 
- to generate maps to visualize your neighborhoods and how they cluster together

Lets get the co-ordinates of Toronto, Ontario

In [21]:
#!pip install geopy

In [22]:
import geocoder
from geopy.geocoders import Nominatim 

address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('These are the coordinates of Toronto {}, {}.'.format(latitude,longitude))

These are the coordinates of Toronto 43.6534817, -79.3839347.


## Now we will create the map of Toronto

In [24]:
#!pip install folium

In [25]:
import folium

In [89]:
# Map of Toronto
map_Toronto = folium.Map(location=[latitude, longitude],zoom_start=12)

# Adding Markers
for latitude, longitude, borough, neighbourhood in zip(combined_data['Latitude'],combined_data['Longitude'],
                                                      combined_data['Borough'],combined_data['Neighbourhood']):
    label = '{}, {}'.format(neighbourhood,borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
    [latitude,longitude],
    radius=5,
    popup=label,
    color='green',
    fill=True).add_to(map_Toronto)

map_Toronto

Initilizing Foursquare API credientials

In [27]:
CLIENT_ID = 'CPU0ZRB5X0E2P0ZNNCCORUDHHU0P2ZZ3OLEQAWP5J2GCPIDY' 
CLIENT_SECRET = '2ZMGRNSU0ENK5UWVMR30HSSR5K1RLXCEVW0GJBKFW2EHJHT1'
VERSION = '20180604' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: CPU0ZRB5X0E2P0ZNNCCORUDHHU0P2ZZ3OLEQAWP5J2GCPIDY
CLIENT_SECRET:2ZMGRNSU0ENK5UWVMR30HSSR5K1RLXCEVW0GJBKFW2EHJHT1


Creating a function to get all the venues categories in Toronto

Seeing the venues in Toronto for each Neighborhood

In [90]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius
            )
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']

        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Category']
    
    return(nearby_venues)

In [91]:
venues_in_toronto = getNearbyVenues(combined_data['Neighbourhood'], combined_data['Latitude'], combined_data['Longitude'])


Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Ontario Provincial Government
Islington Avenue
Malvern, Rouge
Don Mills North
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills South
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
The Danforth  East
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmount Park
Bayview Village
Downsview East
The Danforth

In [92]:
venues_in_toronto.shape

(1342, 5)

So we have 1342 rows and 5 columns in the data, lets look at our data

In [93]:
venues_in_toronto.head()

Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0         Parkwoods               43.753259               -79.329656   
1         Parkwoods               43.753259               -79.329656   
2         Parkwoods               43.753259               -79.329656   
3  Victoria Village               43.725882               -79.315572   
4  Victoria Village               43.725882               -79.315572   

                    Venue        Venue Category  
0         Brookbanks Park                  Park  
1                     KFC  Fast Food Restaurant  
2           Variety Store     Food & Drink Shop  
3  Victoria Village Arena          Hockey Arena  
4             Tim Hortons           Coffee Shop

Lets check the venues by Neighborhood

In [94]:
venues_in_toronto.groupby('Neighbourhood').head()

Neighbourhood  \
0                                             Parkwoods   
1                                             Parkwoods   
2                                             Parkwoods   
3                                      Victoria Village   
4                                      Victoria Village   
...                                                 ...   
1326  Mimico NW, The Queensway West, South of Bloor,...   
1327  Mimico NW, The Queensway West, South of Bloor,...   
1328  Mimico NW, The Queensway West, South of Bloor,...   
1329  Mimico NW, The Queensway West, South of Bloor,...   
1330  Mimico NW, The Queensway West, South of Bloor,...   

      Neighbourhood Latitude  Neighbourhood Longitude                   Venue  \
0                  43.753259               -79.329656         Brookbanks Park   
1                  43.753259               -79.329656                     KFC   
2                  43.753259               -79.329656           Variety Store   
3                  43.725882               -79.315572  Victoria Village Arena   
4                  43.725882               -79.315572             Tim Hortons   
...                      ...                      ...                     ...   
1326               43.628841               -79.520999        South St. Burger   
1327               43.628841               -79.520999              Wingporium   
1328               43.628841               -79.520999               Dollarama   
1329               43.628841               -79.520999          Healthy Planet   
1330               43.628841               -79.520999    Artisano Bakery Café   

            Venue Category  
0                     Park  
1     Fast Food Restaurant  
2        Food & Drink Shop  
3             Hockey Arena  
4              Coffee Shop  
...                    ...  
1326          Burger Joint  
1327           Wings Joint  
1328        Discount Store  
1329       Supplement Shop  
1330                Bakery  

[430 rows x 5 columns]

In [95]:
venues_in_toronto.groupby('Venue Category').max()

Neighbourhood  \
Venue Category                                                          
Accessories Store                    Lawrence Manor, Lawrence Heights   
Adult Boutique                                   Church and Wellesley   
Airport                                                Downsview East   
Airport Food Court  CN Tower, King and Spadina, Railway Lands, Har...   
Airport Gate        CN Tower, King and Spadina, Railway Lands, Har...   
...                                                               ...   
Wine Bar                                     Little Portugal, Trinity   
Wine Shop                                Dufferin, Dovercourt Village   
Wings Joint         Mimico NW, The Queensway West, South of Bloor,...   
Women's Store                                     Caledonia-Fairbanks   
Yoga Studio                            University of Toronto, Harbord   

                    Neighbourhood Latitude  Neighbourhood Longitude  \
Venue Category                                                        
Accessories Store                43.718518               -79.464763   
Adult Boutique                   43.665860               -79.383160   
Airport                          43.737473               -79.394420   
Airport Food Court               43.628947               -79.394420   
Airport Gate                     43.628947               -79.394420   
...                                    ...                      ...   
Wine Bar                         43.653206               -79.400049   
Wine Shop                        43.669005               -79.442259   
Wings Joint                      43.628841               -79.520999   
Women's Store                    43.689026               -79.453512   
Yoga Studio                      43.715383               -79.340923   

                                              Venue  
Venue Category                                       
Accessories Store               Ardene Shoes Outlet  
Adult Boutique                            Seduction  
Airport             Toronto Downsview Airport (YZD)  
Airport Food Court                Billy Bishop Café  
Airport Gate                                 Gate 8  
...                                             ...  
Wine Bar                            Paris Paris Bar  
Wine Shop                                    Macedo  
Wings Joint                              Wingporium  
Women's Store                         Maximum Woman  
Yoga Studio                               YogaSpace  

[244 rows x 4 columns]

In [96]:
venues_in_toronto.groupby('Venue Category').count()

Neighbourhood  Neighbourhood Latitude  \
Venue Category                                              
Accessories Store               1                       1   
Adult Boutique                  1                       1   
Airport                         2                       2   
Airport Food Court              1                       1   
Airport Gate                    1                       1   
...                           ...                     ...   
Wine Bar                        2                       2   
Wine Shop                       1                       1   
Wings Joint                     1                       1   
Women's Store                   1                       1   
Yoga Studio                     9                       9   

                    Neighbourhood Longitude  Venue  
Venue Category                                      
Accessories Store                         1      1  
Adult Boutique                            1      1  
Airport                                   2      2  
Airport Food Court                        1      1  
Airport Gate                              1      1  
...                                     ...    ...  
Wine Bar                                  2      2  
Wine Shop                                 1      1  
Wings Joint                               1      1  
Women's Store                             1      1  
Yoga Studio                               9      9  

[244 rows x 4 columns]

So there are around 430 Neighborhood and 244 different Venue Categories

## We will use One hot encoder on Venue Category 

In [97]:
toronto_venue_category = pd.get_dummies(venues_in_toronto[['Venue Category']],prefix="", prefix_sep="")

# Moving Neighbourhood column in the beginnign
toronto_venue_category['Neighbourhood'] = venues_in_toronto['Neighbourhood']
fixed_column = [toronto_venue_category.columns[-1]] + list(toronto_venue_category.columns[:-1])
toronto_venue_category = toronto_venue_category[fixed_column]

toronto_venue_category.head()


Neighbourhood  Accessories Store  Adult Boutique  Airport  \
0         Parkwoods                  0               0        0   
1         Parkwoods                  0               0        0   
2         Parkwoods                  0               0        0   
3  Victoria Village                  0               0        0   
4  Victoria Village                  0               0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  American Restaurant  ...  Vegetarian / Vegan Restaurant  \
0                 0                    0  ...                              0   
1                 0                    0  ...                              0   
2                 0                    0  ...                              0   
3                 0                    0  ...                              0   
4                 0                    0  ...                              0   

   Video Game Store  Video Store  Vietnamese Restaurant  Warehouse Store  \
0                 0            0                      0                0   
1                 0            0                      0                0   
2                 0            0                      0                0   
3                 0            0                      0                0   
4                 0            0                      0                0   

   Wine Bar  Wine Shop  Wings Joint  Women's Store  Yoga Studio  
0         0          0            0              0            0  
1         0          0            0              0            0  
2         0          0            0              0            0  
3         0          0            0              0            0  
4         0          0            0              0            0  

[5 rows x 245 columns]

We will group our data by Neighbourhoods, calculate the mean of Venue category in each neighbourhood

In [98]:
toronto_grouped = toronto_venue_category.groupby('Neighbourhood').mean().reset_index()
toronto_grouped.head()

Neighbourhood  Accessories Store  \
0                                        Agincourt                0.0   
1                           Alderwood, Long Branch                0.0   
2  Bathurst Manor, Wilson Heights, Downsview North                0.0   
3                                  Bayview Village                0.0   
4                Bedford Park, Lawrence Manor East                0.0   

   Adult Boutique  Airport  Airport Food Court  Airport Gate  Airport Lounge  \
0             0.0      0.0                 0.0           0.0             0.0   
1             0.0      0.0                 0.0           0.0             0.0   
2             0.0      0.0                 0.0           0.0             0.0   
3             0.0      0.0                 0.0           0.0             0.0   
4             0.0      0.0                 0.0           0.0             0.0   

   Airport Service  Airport Terminal  American Restaurant  ...  \
0              0.0               0.0                 0.00  ...   
1              0.0               0.0                 0.00  ...   
2              0.0               0.0                 0.00  ...   
3              0.0               0.0                 0.00  ...   
4              0.0               0.0                 0.04  ...   

   Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
0                            0.0               0.0          0.0   
1                            0.0               0.0          0.0   
2                            0.0               0.0          0.0   
3                            0.0               0.0          0.0   
4                            0.0               0.0          0.0   

   Vietnamese Restaurant  Warehouse Store  Wine Bar  Wine Shop  Wings Joint  \
0                    0.0              0.0       0.0        0.0          0.0   
1                    0.0              0.0       0.0        0.0          0.0   
2                    0.0              0.0       0.0        0.0          0.0   
3                    0.0              0.0       0.0        0.0          0.0   
4                    0.0              0.0       0.0        0.0          0.0   

   Women's Store  Yoga Studio  
0            0.0          0.0  
1            0.0          0.0  
2            0.0          0.0  
3            0.0          0.0  
4            0.0          0.0  

[5 rows x 245 columns]

Lets make a function to get the top common venue category

In [99]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

We will take 15 top to cluster our neighbourhood

In [100]:
num_top_venues = 15

indicators = ['st', 'nd', 'rd']

# create columns accordingly to the number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))


# Lets Create a new Dataframe
n_venue_sorted = pd.DataFrame(columns=columns)
n_venue_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    n_venue_sorted.iloc[ind,1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

n_venue_sorted.head()

Neighbourhood 1st Most Common Venue  \
0                                        Agincourt        Breakfast Spot   
1                           Alderwood, Long Branch           Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North                  Bank   
3                                  Bayview Village                  Café   
4                Bedford Park, Lawrence Manor East           Coffee Shop   

  2nd Most Common Venue      3rd Most Common Venue 4th Most Common Venue  \
0                Lounge  Latin American Restaurant     Accessories Store   
1            Playground             Sandwich Place                   Gym   
2           Coffee Shop                Pizza Place         Shopping Mall   
3   Japanese Restaurant         Chinese Restaurant                  Bank   
4        Sandwich Place         Italian Restaurant      Toy / Game Store   

  5th Most Common Venue 6th Most Common Venue      7th Most Common Venue  \
0          Neighborhood    Mexican Restaurant  Middle Eastern Restaurant   
1                   Pub              Pharmacy                Coffee Shop   
2   Fried Chicken Joint    Frozen Yogurt Shop             Sandwich Place   
3     Accessories Store                Museum               Neighborhood   
4        Breakfast Spot               Butcher                       Café   

  8th Most Common Venue 9th Most Common Venue      10th Most Common Venue  \
0    Miscellaneous Shop     Mobile Phone Shop  Modern European Restaurant   
1    Athletics & Sports   Monument / Landmark                      Museum   
2           Supermarket    Chinese Restaurant   Middle Eastern Restaurant   
3           Music Venue         Movie Theater                   Nightclub   
4      Sushi Restaurant          Liquor Store             Thai Restaurant   

  11th Most Common Venue 12th Most Common Venue      13th Most Common Venue  \
0    Monument / Landmark                  Motel               Movie Theater   
1          Movie Theater                  Motel          Miscellaneous Shop   
2            Gas Station             Restaurant                   Gift Shop   
3                  Motel    Monument / Landmark  Modern European Restaurant   
4   Fast Food Restaurant             Restaurant                   Juice Bar   

       14th Most Common Venue 15th Most Common Venue  
0                      Museum            Music Venue  
1  Modern European Restaurant      Mobile Phone Shop  
2                       Diner                   Park  
3           Mobile Phone Shop     Miscellaneous Shop  
4     Comfort Food Restaurant      Indian Restaurant

## Let's start making Clusters

In [101]:
from sklearn.cluster import KMeans

In [102]:
# Setting the custer number to 5

k_num_cluster = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

# running k means
kmeans = KMeans(n_clusters=k_num_cluster, random_state=0).fit(toronto_grouped_clustering)
kmeans

KMeans(n_clusters=5, random_state=0)

In [103]:
# Lets view the labels
kmeans.labels_

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 3, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0,
       2, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 2, 0, 0, 0,
       1, 0, 0, 1, 0, 0, 0, 0, 4, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 2, 0, 0,
       0, 0, 0, 0, 3, 0, 0, 0, 3, 0, 0, 1])

In [104]:
# Adding the labels to the top 15 common venue categories
n_venue_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

Lets Join Toronto_grouped with Combined data, so we can plot the data

In [107]:
toronto_merged = combined_data

toronto_merged = toronto_merged.join(n_venue_sorted.set_index('Neighbourhood'), on='Neighbourhood')

toronto_merged.head()

PostalCode           Borough                     Neighbourhood   Latitude  \
0        M3A        North York                         Parkwoods  43.753259   
1        M4A        North York                  Victoria Village  43.725882   
2        M5A  Downtown Toronto         Regent Park, Harbourfront  43.654260   
3        M6A        North York  Lawrence Manor, Lawrence Heights  43.718518   
4        M7A      Queen's Park     Ontario Provincial Government  43.662301   

   Longitude  Cluster Labels  1st Most Common Venue   2nd Most Common Venue  \
0 -79.329656             1.0   Fast Food Restaurant       Food & Drink Shop   
1 -79.315572             0.0  Portuguese Restaurant             Pizza Place   
2 -79.360636             0.0            Coffee Shop                    Park   
3 -79.464763             0.0         Clothing Store  Furniture / Home Store   
4 -79.389494             0.0            Coffee Shop        Sushi Restaurant   

  3rd Most Common Venue  4th Most Common Venue  ... 6th Most Common Venue  \
0                  Park            Music Venue  ...           Men's Store   
1          Hockey Arena            Coffee Shop  ...     Accessories Store   
2                Bakery                Theater  ...           Yoga Studio   
3     Accessories Store  Vietnamese Restaurant  ...           Coffee Shop   
4           Yoga Studio               Creperie  ...         Burrito Place   

        7th Most Common Venue  8th Most Common Venue  \
0               Metro Station     Mexican Restaurant   
1  Modern European Restaurant                 Museum   
2                         Pub  Performing Arts Venue   
3                    Boutique               Pharmacy   
4               Smoothie Shop                   Café   

       9th Most Common Venue 10th Most Common Venue  \
0  Middle Eastern Restaurant     Miscellaneous Shop   
1              Movie Theater                  Motel   
2               Dessert Shop             Restaurant   
3                  Pet Store     Mexican Restaurant   
4             Sandwich Place     College Auditorium   

      11th Most Common Venue      12th Most Common Venue  \
0          Mobile Phone Shop  Modern European Restaurant   
1        Monument / Landmark          Miscellaneous Shop   
2        Distribution Center              Chocolate Shop   
3  Middle Eastern Restaurant          Miscellaneous Shop   
4         Persian Restaurant                         Bar   

  13th Most Common Venue 14th Most Common Venue      15th Most Common Venue  
0    Monument / Landmark                  Motel               Movie Theater  
1      Mobile Phone Shop           Neighborhood   Middle Eastern Restaurant  
2                   Café         Farmers Market                         Spa  
3             Playground      Mobile Phone Shop  Modern European Restaurant  
4                  Diner    Distribution Center                        Park  

[5 rows x 21 columns]

In [108]:
# Lets drop NaN values (if any)

toronto_merged_nonan = toronto_merged.dropna(subset=['Cluster Labels'])

Plotting our clusters

In [109]:
import matplotlib.cm as cm
import matplotlib.colors as colors

In [110]:
map_clusters = folium.Map(location=[latitude, longitude],zoom_start=11)

# Setting colors
x = np.arange(k_num_cluster)
ys = [i + x + (i*x)**2 for i in range(k_num_cluster)]
colors_array = cm.rainbow(np.linspace(0,1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged_nonan['Latitude'],toronto_merged_nonan['Longitude'],
                                  toronto_merged_nonan['Neighbourhood'],toronto_merged_nonan['Cluster Labels']):
    label=folium.Popup('Cluster '+ str(int(cluster) +1) + '\n' + str(poi), parse_html=True)
    folium.CircleMarker([lat, lon],
                       radius=5,
                       popup=label,
                       color=rainbow[int(cluster-1)],
                       fill=True,
                       fill_color=rainbow[int(cluster-1)]).add_to(map_clusters)
map_clusters

Lets Verify our Clusters

### Cluster 1

In [112]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 0, toronto_merged_nonan.columns[[1]+list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels  1st Most Common Venue  \
1               North York             0.0  Portuguese Restaurant   
2         Downtown Toronto             0.0            Coffee Shop   
3               North York             0.0         Clothing Store   
4             Queen's Park             0.0            Coffee Shop   
6              Scarborough             0.0   Fast Food Restaurant   
..                     ...             ...                    ...   
96        Downtown Toronto             0.0                   Café   
97        Downtown Toronto             0.0                   Café   
99        Downtown Toronto             0.0                    Pub   
100  East Toronto Business             0.0     Light Rail Station   
102              Etobicoke             0.0                 Bakery   

      2nd Most Common Venue 3rd Most Common Venue  4th Most Common Venue  \
1               Pizza Place          Hockey Arena            Coffee Shop   
2                      Park                Bakery                Theater   
3    Furniture / Home Store     Accessories Store  Vietnamese Restaurant   
4          Sushi Restaurant           Yoga Studio               Creperie   
6         Accessories Store           Men's Store          Metro Station   
..                      ...                   ...                    ...   
96       Italian Restaurant                Bakery             Restaurant   
97              Coffee Shop            Restaurant                  Hotel   
99                 Beer Bar   Martial Arts School            Men's Store   
100                    Park                Garden          Garden Center   
102  Thrift / Vintage Store           Social Club   Fast Food Restaurant   

    5th Most Common Venue      6th Most Common Venue  \
1       French Restaurant          Accessories Store   
2          Breakfast Spot                Yoga Studio   
3             Event Space                Coffee Shop   
4                     Spa              Burrito Place   
6      Mexican Restaurant  Middle Eastern Restaurant   
..                    ...                        ...   
96                    Pub                       Park   
97     Seafood Restaurant                Pizza Place   
99              Bookstore         Mexican Restaurant   
100               Brewery              Auto Workshop   
102           Flower Shop          Convenience Store   

          7th Most Common Venue  8th Most Common Venue  \
1    Modern European Restaurant                 Museum   
2                           Pub  Performing Arts Venue   
3                      Boutique               Pharmacy   
4                 Smoothie Shop                   Café   
6            Miscellaneous Shop      Mobile Phone Shop   
..                          ...                    ...   
96                      Butcher   Caribbean Restaurant   
97                          Pub              Speakeasy   
99              Bubble Tea Shop           Burger Joint   
100                  Skate Park             Restaurant   
102             Supplement Shop         Sandwich Place   

          9th Most Common Venue 10th Most Common Venue  \
1                 Movie Theater                  Motel   
2                  Dessert Shop             Restaurant   
3                     Pet Store     Mexican Restaurant   
4                Sandwich Place     College Auditorium   
6    Modern European Restaurant    Monument / Landmark   
..                          ...                    ...   
96                   Playground                 Market   
97                   Steakhouse               Tea Room   
99                         Café   Ethiopian Restaurant   
100                         Spa          Burrito Place   
102               Tanning Salon             Kids Store   

        11th Most Common Venue  12th Most Common Venue 13th Most Common Venue  \
1          Monument / Landmark      Miscellaneous Shop      Mobile Phone Shop   
2          Distribution Center          Chocolat

### Cluster 2

In [113]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 1, toronto_merged_nonan.columns[[1]+list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0               North York             1.0  Fast Food Restaurant   
21                    York             1.0                  Park   
35  East York/East Toronto             1.0     Convenience Store   
49              North York             1.0      Basketball Court   
61         Central Toronto             1.0    Photography Studio   
66              North York             1.0     Convenience Store   
68         Central Toronto             1.0         Jewelry Store   
77               Etobicoke             1.0     Mobile Phone Shop   
83         Central Toronto             1.0                Lawyer   
85             Scarborough             1.0            Playground   
91        Downtown Toronto             1.0                  Park   

         2nd Most Common Venue 3rd Most Common Venue    4th Most Common Venue  \
0            Food & Drink Shop                  Park              Music Venue   
21               Women's Store                   Bar        Accessories Store   
35                        Park     Accessories Store  New American Restaurant   
49  Construction & Landscaping                Bakery                     Park   
61                 Swim School                  Park                 Bus Line   
66           Electronics Store                  Park        Accessories Store   
68                        Park                 Trail         Sushi Restaurant   
77                        Park        Sandwich Place        Accessories Store   
83                        Park          Tennis Court        Accessories Store   
85                Intersection                  Park        Accessories Store   
91                  Playground                 Trail              Music Venue   

         5th Most Common Venue 6th Most Common Venue  \
0     Mediterranean Restaurant           Men's Store   
21  Modern European Restaurant                Museum   
35               Metro Station    Mexican Restaurant   
49           Accessories Store                Museum   
61           Accessories Store           Music Venue   
66                Neighborhood         Metro Station   
68                      Museum         Movie Theater   
77                Neighborhood           Men's Store   
83                 Music Venue           Men's Store   
85  Modern European Restaurant                Museum   
91    Mediterranean Restaurant           Men's Store   

        7th Most Common Venue      8th Most Common Venue  \
0               Metro Station         Mexican Restaurant   
21              Movie Theater                      Motel   
35  Middle Eastern Restaurant         Miscellaneous Shop   
49              Movie Theater                      Motel   
61              Metro Station         Mexican Restaurant   
66         Mexican Restaurant  Middle Eastern Restaurant   
68                      Motel        Monument / Landmark   
77              Metro Station         Mexican Restaurant   
83              Metro Station         Mexican Restaurant   
85              Movie Theater                      Motel   
91              Metro Station         Mexican Restaurant   

         9th Most Common Venue      10th Most Common Venue  \
0    Middle Eastern Restaurant          Miscellaneous Shop   
21         Monument / Landmark           Mobile Phone Shop   
35           Mobile Phone Shop  Modern European Restaurant   
49         Monument / Landmark  Modern European Restaurant   
61   Middle Eastern Restaurant          Miscellaneous Shop   
66          Miscellaneous Shop           Mobile Phone Shop   
68  Modern European Restaurant           Accessories Store   
77   Middle Eastern Restaurant          Miscellaneous Shop   
83   Middle Eastern Restaurant          Miscellaneous Shop   
85         Monument / Landmark           Mobile Phone Shop   
91   Middle Eastern Restaurant          Miscellaneous Shop   

        11th Most Common Venue      12th Most Common Venue  \
0            Mobile Phone Sh

### Cluster 3

In [114]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 2, toronto_merged_nonan.columns[[1]+list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
57   North York             2.0        Baseball Field   
98    Etobicoke             2.0                  Pool   
101   Etobicoke             2.0                  Pool   

          2nd Most Common Venue 3rd Most Common Venue  \
57            Accessories Store          Neighborhood   
98                        River     Accessories Store   
101  Construction & Landscaping        Baseball Field   

       4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
57               Men's Store         Metro Station    Mexican Restaurant   
98              Neighborhood           Men's Store         Metro Station   
101  New American Restaurant         Metro Station    Mexican Restaurant   

         7th Most Common Venue      8th Most Common Venue  \
57   Middle Eastern Restaurant         Miscellaneous Shop   
98          Mexican Restaurant  Middle Eastern Restaurant   
101  Middle Eastern Restaurant         Miscellaneous Shop   

    9th Most Common Venue      10th Most Common Venue  \
57      Mobile Phone Shop  Modern European Restaurant   
98     Miscellaneous Shop           Mobile Phone Shop   
101     Mobile Phone Shop  Modern European Restaurant   

         11th Most Common Venue 12th Most Common Venue 13th Most Common Venue  \
57          Monument / Landmark                  Motel          Movie Theater   
98   Modern European Restaurant    Monument / Landmark                  Motel   
101         Monument / Landmark                  Motel          Movie Theater   

    14th Most Common Venue 15th Most Common Venue  
57                  Museum            Music Venue  
98           Movie Theater                 Museum  
101                 Museum            Music Venue

### Cluster 4

In [115]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 3, toronto_merged_nonan.columns[[1]+list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
40  North York             3.0               Airport                  Park   
52  North York             3.0                  Park     Accessories Store   
64        York             3.0                  Park     Accessories Store   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
40     Accessories Store          Neighborhood           Men's Store   
52          Neighborhood           Men's Store         Metro Station   
64          Neighborhood           Men's Store         Metro Station   

   6th Most Common Venue      7th Most Common Venue  \
40         Metro Station         Mexican Restaurant   
52    Mexican Restaurant  Middle Eastern Restaurant   
64    Mexican Restaurant  Middle Eastern Restaurant   

        8th Most Common Venue 9th Most Common Venue  \
40  Middle Eastern Restaurant    Miscellaneous Shop   
52         Miscellaneous Shop     Mobile Phone Shop   
64         Miscellaneous Shop     Mobile Phone Shop   

        10th Most Common Venue      11th Most Common Venue  \
40           Mobile Phone Shop  Modern European Restaurant   
52  Modern European Restaurant         Monument / Landmark   
64  Modern European Restaurant         Monument / Landmark   

   12th Most Common Venue 13th Most Common Venue 14th Most Common Venue  \
40    Monument / Landmark                  Motel          Movie Theater   
52                  Motel          Movie Theater                 Museum   
64                  Motel          Movie Theater                 Museum   

   15th Most Common Venue  
40                 Museum  
52            Music Venue  
64            Music Venue

### Cluster 5

In [117]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 4, toronto_merged_nonan.columns[[1]+list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
32  Scarborough             4.0            Playground     Accessories Store   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
32          Neighborhood           Men's Store         Metro Station   

   6th Most Common Venue      7th Most Common Venue 8th Most Common Venue  \
32    Mexican Restaurant  Middle Eastern Restaurant    Miscellaneous Shop   

   9th Most Common Venue      10th Most Common Venue 11th Most Common Venue  \
32     Mobile Phone Shop  Modern European Restaurant    Monument / Landmark   

   12th Most Common Venue 13th Most Common Venue 14th Most Common Venue  \
32                  Motel          Movie Theater                 Museum   

   15th Most Common Venue  
32            Music Venue

We have sucessfully Clustered our data into 5 Clusters